File: src/main/java/net/sf/rails/game/state/GenericStateChange.java
package net.sf.rails.game.state;

public final class GenericStateChange<E> extends Change {

    final private GenericState<E> state;
    final private E previous, next;

    public GenericStateChange(GenericState<E> state, E object) {
        this.state = state;
        previous = state.value();
        next = object;
        super.init(state);
    }

    @Override void execute() {
        state.change(next);
    }

    @Override void undo() {
        state.change(previous);
    }

    @Override
    public GenericState<E> getState() {
        return state;
    }

    @Override
    public String toString() {
        return "Change for " + state + ": Replace " + previous + " by " + next ;
    }

}

File: dotnet/RailsLib.Net/Net/Game/State/GenericStateChange.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace GameLib.Net.Game.State
{
    public sealed class GenericStateChange<T> : Change
    {
        private readonly GenericState<T> state;
        private readonly T previous, next;

        public GenericStateChange(GenericState<T> state, T o)
        {
            this.state = state;
            previous = state.Value;
            next = o;
            base.Init(state);
        }

        override public void Execute()
        {
            state.Change(next);
        }

        override public void Undo()
        {
            state.Change(previous);
        }

        public override GameState GameState
        {
            get
            {
                return state;
            }
        }

        override public string ToString()
        {
            return "Change for " + state + ": Replace " + previous + " by " + next;
        }
    }
}
